haskell/lint.bzl

```diff
@@ -0,0 +1,294 @@
+"""Lint Haskell sources using hlint."""
+
+load("@bazel_skylib//lib:dicts.bzl", "dicts")
+load("@bazel_skylib//lib:paths.bzl", "paths")
+load(
+    "@io_tweag_rules_haskell//haskell:private/path_utils.bzl",
+    "match_label",
+    "parse_pattern",
+)
+load("@io_tweag_rules_haskell//haskell:providers.bzl", "HaskellInfo")
+
+HaskellHLintInfo = provider(
+    doc = "Provider that collects files produced by hlint",
+    fields = {
+        "outputs": "dict from target Label to hlint results.",
+    },
+)
+
+def _collect_hlint_logs(deps):
+    return dicts.add(*[
+        dep[HaskellHLintInfo].outputs
+        for dep in deps
+        if HaskellHLintInfo in dep
+    ])
+
+def _haskell_lint_aspect_impl(target, ctx):
+    if not HaskellInfo in target:
+        return []
+
+    hlint_toolchain = ctx.toolchains["@io_tweag_rules_haskell//haskell/lint:toolchain"]
+    hint = hlint_toolchain.hint
+
+    inputFiles = [
+        f
+        for src in getattr(ctx.rule.attr, "srcs", [])
+        for f in src.files
+        if f.is_source and f.extension in ["chs", "hs", "hs-boot", "hsc", "lhs", "lhs-boot"]
```

Comment:
HLint seems to croak on `.chs` files that actually use C2hs syntax. I'm not sure we can lint those.

Reply:
Good point, thanks fixed.